rn/Teacher/ios/Teacher/SpeedGrader/URLSubmissionViewer.swift

```diff
@@ -0,0 +1,85 @@
+//
+// This file is part of Canvas.
+// Copyright (C) 2020-present  Instructure, Inc.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+//
+
+import SwiftUI
+import Core
+
+struct URLSubmissionViewer: View {
+    let submission: Submission
+
+    @Environment(\.appEnvironment) var env
+    @Environment(\.viewController) var controller
+
+    @State var image: UIImage?
+    @State var loaded: URL?
+    @State var error: Error?
+
+    var body: some View {
+        ScrollView {
+            VStack(spacing: 0) {
+                HStack {
+                    Text("This submission is a URL to an external page. We've included a snapshot of what the page looked like when it was submitted.")
+                        .font(.regular16).foregroundColor(.textDarkest)
+                    Spacer()
+                }
+                Button(action: {
+                    guard let url = submission.url else { return }
+                    env.loginDelegate?.openExternalURL(url)
+                }, label: {
+                    Text(submission.url?.absoluteString ?? "")
+                        .font(.semibold16).foregroundColor(Color(Brand.shared.linkColor))
+                        .padding(.vertical, 16)
+                })
+                if let file = submission.attachments?.first, let url = file.url {
+                    if let error = error, url == loaded {
+                        Text(error.localizedDescription)
+                            .font(.regular16).foregroundColor(.textDanger)
+                    } else if let image = image, url == loaded {
+                        Button(action: {
+                            guard let id = submission.attachments?.first?.id else { return }
```

Comment:
Is there harm in using `file` (from line 48 above) here instead of looking inside `submission.attachments?.first` again?

Reply:
nope. I can fix that in the next PR.